### PR TITLE
Prevents Locale Code build failures #2967

### LINKED ISSUE #2967: [BUG] Github actions test causes failure on certain local code -Dtests.locale=az-Cyrl
**What is the bug?**
During a [GH action from a commit to a PR](https://github.com/opensearch-project/ml-commons/actions/runs/10908291086/job/30281922405) I had failures I one by one took the reproduction lines [for example](https://github.com/opensearch-project/ml-commons/actions/runs/10908291086/job/30281922405#step:6:2144) Takes you to this command `./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testPrepareRequest" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21` However when you change the `Dtests.locale=az-Cyrl` to `Dtests.locale=hsb ` or `Dtests.locale=en-US ` it ends up passing

**How can one reproduce the bug?**
1. Go to your terminal that has the ml-commons repo for example from main
2. You can run all commands in one sitting or test one by one however All will fail 
```
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testPrepareRequest" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest_LocalModelInferenceDisabled" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testPrepareBatchRequest" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest_RemoteInferenceDisabled" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest_LocalModelInferenceDisabled" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testPrepareBatchRequest" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
./gradlew ':opensearch-ml-plugin:test' --tests "org.opensearch.ml.rest.RestMLPredictionActionTests.testGetRequest_RemoteInferenceDisabled" -Dtests.seed=1FD264282B2B1F19 -Dtests.security.manager=false -Dtests.locale=az-Cyrl -Dtests.timezone=Pacific/Tarawa -Druntime.java=21
```
3. Now you can you can see that if you change `az-Cyrl` to (`en-US` or `hsb`) they will all pass

**What is the expected behavior?**
Normally running these gradle commands should result in build succeeded however they all evaluate to failures only when you change them to hsb they will pass. Also github actions should be configured to work with valid local codes or ml-commons should onboard az-Cyrl

**What is your host/environment?**
 - OS: MACOS 14.6.1 
 - OpenSearch Version: 2.16
 - Plugins: ml-commons

**Do you have any screenshots?**
I have a video showing I can forward on request

**Do you have any additional context?**
I was on this commit id when I did the experiment 0d2693180bbd76aab7048490ce9eb4dc5eeb3fe1 Thanks for @b4sjoo Helping debug this problem

### PR DESCRIPTION
**Pull request by brianf-aws**
### Description
Depending on the GitHub CI workflow executes it will invoke ./gradlew build which will run tests on random parameters every time, there are locale-code's that will break a build one of them being **az-Cyrl**. The solution was to update the codebase that involved string operations such as `toUpperCase(), toLowerCase()` to `toUpperCase(Locale.Root), toLowerCase(Locale.Root)` In doing this we make the string operations language agnostic meaning it wont be affected on a computer that is set to a different language like Spanish or Turkish.  

#### Testing
- `./gradlew build -DTests.Locale=az-Cyrl` which previously failed now works
- `./gradlew build` builds successfully too.

**TLDR:**
- The change was to update String operations to use `Locale.Root` as a argument to avoid default Locale (which may depend on one that is not supported) you can see this [here](https://docs.oracle.com/javase/8/docs/api/java/util/Locale.html#getDefault--) . 

### Related Issues
Resolves #2967

### Check List
- [X] New functionality includes testing.
- [X] Commits are signed per the DCO using `--signoff`.
- [X] Public documentation issue/PR [created](https://github.com/opensearch-project/documentation-website/issues/new/choose).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/ml-commons/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

**Review by dbwiddis — CHANGES_REQUESTED** [review]
This seems like the wrong fix for the issue.

There's clearly a locale-based bug somewhere but instead of fixing it you're ignoring it for the entire build, even for tests unaffected by the issue.  This would miss issues such as https://github.com/opensearch-project/ml-commons/pull/1734

All of the failing tests in the linked issue are in the `RestMLPredictionActionTests` class, so if those tests require `en-US` locale that should be scoped only to those tests.  

However, locale based bugs are very frequently using upper- or lower-case string transformations.   This line has a `toUpperCase()` transformation without a locale and is very likely the bug that needs fixing by adding `Locale.ROOT`.  (There may be others.)

https://github.com/opensearch-project/ml-commons/blob/6a6cac13ee1a009ea43a3bad147059578d2adc38/plugin/src/main/java/org/opensearch/ml/rest/RestMLPredictionAction.java#L132

A brief search of the entire repo shows several similar potential locale issues, you should probably fix them all!
 - https://github.com/search?q=repo%3Aopensearch-project%2Fml-commons%20touppercase()&type=code
 - https://github.com/search?q=repo%3Aopensearch-project%2Fml-commons+tolowercase%28%29&type=code

If you want to go a step further, converting a string by upper/lower casing just for a comparison is inefficient as it creates a whole new string object which is immediately discarded.  If done frequently this adds to the heap/GC load.  Use `.equalsIgnoreCase()` when you can to avoid this extra object creation.

**Comment by dbwiddis** [comment]
Update: this is the specific bug location, but as noted above you should fix all of them.
https://github.com/opensearch-project/ml-commons/blob/6a6cac13ee1a009ea43a3bad147059578d2adc38/common/src/main/java/org/opensearch/ml/common/connector/ConnectorAction.java#L211

**Comment by brianf-aws** [comment]
> This seems like the wrong fix for the issue.
> 
> There's clearly a locale-based bug somewhere but instead of fixing it you're ignoring it for the entire build, even for tests unaffected by the issue. This would miss issues such as #1734
> 
> All of the failing tests in the linked issue are in the `RestMLPredictionActionTests` class, so if those tests require `en-US` locale that should be scoped only to those tests.
> 
> However, locale based bugs are very frequently using upper- or lower-case string transformations. This line has a `toUpperCase()` transformation without a locale and is very likely the bug that needs fixing by adding `Locale.ROOT`. (There may be others.)
> 
> https://github.com/opensearch-project/ml-commons/blob/6a6cac13ee1a009ea43a3bad147059578d2adc38/plugin/src/main/java/org/opensearch/ml/rest/RestMLPredictionAction.java#L132
> 
> A brief search of the entire repo shows several similar potential locale issues, you should probably fix them all!
> 
> * https://github.com/search?q=repo%3Aopensearch-project%2Fml-commons%20touppercase()&type=code
> * https://github.com/search?q=repo%3Aopensearch-project%2Fml-commons+tolowercase%28%29&type=code
> 
> If you want to go a step further, converting a string by upper/lower casing just for a comparison is inefficient as it creates a whole new string object which is immediately discarded. If done frequently this adds to the heap/GC load. Use `.equalsIgnoreCase()` when you can to avoid this extra object creation.

Hey Dan! Thanks for the insight, I agree with you I did feel skeptical making the change Ill get to work with updating the string methods to use `Local.ROOT`

**Review by dhrubo-os — APPROVED** [review]

**Comment by dhrubo-os** [comment]
Thanks for making the changes.

**Review by dbwiddis — APPROVED** [review]
LGTM Someone should probably do an org-wide search for these. :)

(Edit: [just did](https://github.com/search?q=org%3Aopensearch-project+touppercase%28%29&type=code) 172 of them!  Holy cow!)

**Review by rbhavna — APPROVED** [review]

**Comment by ylwu-amzn** [comment]
> LGTM Someone should probably do an org-wide search for these. :)
> 
> (Edit: [just did](https://github.com/search?q=org%3Aopensearch-project+touppercase%28%29&type=code) 172 of them! Holy cow!)

Thanks Dan, can you create issues in the impacted repos ?

**Comment by dbwiddis** [comment]
> > LGTM Someone should probably do an org-wide search for these. :)
> > (Edit: [just did](https://github.com/search?q=org%3Aopensearch-project+touppercase%28%29&type=code) 172 of them! Holy cow!)
> 
> Thanks Dan, can you create issues for the impacted repos ?

Yep, sometime next week.   Filtering to just Java files it's only 66.
